examples/crd_derive_schema.rs

```diff
@@ -0,0 +1,244 @@
+//
+// References:
+// - https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#defaulting
+// - https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#defaulting-and-nullable
+
+#[derive(CustomResource, Serialize, Deserialize, Default, Debug, PartialEq, Clone, JsonSchema)]
+#[kube(
+    group = "clux.dev",
+    version = "v1",
+    kind = "Foo",
+    namespaced,
+    derive = "PartialEq",
+    derive = "Default"
+)]
+#[kube(apiextensions = "v1")]
+pub struct FooSpec {
+    // Non-nullable without default is required.
+    //
+    // There shouldn't be any ambiguity here.
+    non_nullable: String,
+
+    // Non-nullable with default value.
+    //
+    // Serializing will work as expected because the field cannot be `None`.
+    //
+    // When deserializing a response from the server, the field should always be a string because
+    // the field is non-nullable and the server sets the value to the default specified in the schema.
+    //
+    // When deserializing some input, the default value will be set if missing.
+    // However, if `null` is specified, `serde` will panic.
+    // The server prunes `null` for non-nullable field since 1.20 and the default is applied.
+    // To match the server's behavior exactly, we can use a custom deserializer.
+    #[serde(default = "default_value")]
+    non_nullable_with_default: String,
+
+    // Nullable without default, skipping None.
+    //
+    // By skipping to serialize, the field won't be present in the object.
+    // If serialized as `null` (next field), the object will have the field set to `null`.
+    //
+    // Deserializing works as expected either way. `None` if it's missing or `null`.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    nullable_skipped: Option<String>,
+    // Nullable without default, not skipping None.
+    nullable: Option<String>,
+
+    // Nullable with default, skipping None.
+    //
+    // By skipping to serialize when `None`, the server will set the the default value specified in the schema.
+    // If serialized as `null`, the server will conserve it and the defaulting does not happen (since 1.20).
+    //
+    // When deserializing, the default value is used only when it's missing (`null` is `None`).
+    // This is consistent with how the server handles it since 1.20.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(default = "default_nullable")]
+    nullable_skipped_with_default: Option<String>,
+
+    // Nullable with default, not skipping None.
+    //
+    // The default value won't be used unless missing, so this will set the value to `null`.
+    // If the resource is created with `kubectl` and if this field was missing, defaulting will happen.
+    #[serde(default = "default_nullable")]
+    nullable_with_default: Option<String>,
```

Comment:
Could we add a chrono timestamp as well?

```rust
timestamp: DateTime<Utc>,
```

I think this should work fine as long as we enable the `chrono` feature in `schemars` (which we should do because it's a common thing to have timestamps in Status structs), and maybe add `chrono` to examples

Reply:
Should `kube-derive` depend on `schemars`? I wasn't sure what to do so it's currently in `dev-dependency`.

I added to the integration test and it worked with:
```toml
schemars = { version = "0.8.0", features = ["chrono"] }
chrono = "0.4"
```

For the example, adding a timestamp field prevents deriving `Default`, so it's less convenient. I can add it if you want.

Reply:
Ah, good point. It's best if we can push the dependency to the user (so they could select the features they want, and maybe minimize the set). It does look like that is possible because you're only using `schemars` within quoted code that is only executed during the `Foo::crd` call

Reply:
Ah, also true about the `Default` derive, well, I see you have gotten it into one example at least. That's good enough.